### PR TITLE
Handle account data retrieval error

### PR DESCRIPTION
**Pull request by Mayisha**
### Description
When there is an `errors` object returned by Stripe instead of `account` data, a warning message appears on the screen.

undefined capabilities

# Testing instructions
- Easiest way to reproduce the warning would to return `WP_Error` from the `retrieve` method [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/538d9c2dd236f96cb5fdce789fab37bb3d35eef6/includes/class-wc-stripe-api.php#L168)

### CODE REVIEW

File: includes/class-wc-stripe-account.php
Comment (harriswong) [inline includes/class-wc-stripe-account.php]: Does the account array always have a `errors` attribute? ie. could `$account['errors'] = []` or `$account['errors'] = ''`?

Comment (Mayisha) [reply includes/class-wc-stripe-account.php]: When the request is successful there is only `account` data in response and the `errors` attribute is not there. But I think the code should be able to handle the cases you mentioned as well. Fixed in [this commit](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1969/commits/919331815635a31f7a7b66977e08bd8488976812)

Comment (Mayisha) [reply includes/class-wc-stripe-account.php]: On second thought, we should actually avoid caching the error object. It was being cached before. Fixed [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1969/commits/6225152a8e4829378652df1a266f7c99f0de8643)

File: includes/class-wc-stripe-account.php
Comment (harriswong) [inline includes/class-wc-stripe-account.php:83]: @Mayisha , I like the idea of not caching an account that has errored. Looking at the `cache_account()` function here, it seems like we are trying to retrieve account info by calling our Stripe's client (`$this->stripe_api`). Then our Stripe client invokes a call to Stripe's endpoint at `https://api.stripe.com/v1/account` ([doc](https://stripe.com/docs/api/accounts/retrieve)).

To test, I updated my test key to an invalid key so that the account API will return an error. 

The API response will come back with a JSON error object. 
```
{
 error: {
    message: "...",
    type: "..."
 }
}
```
image

Now, when it goes into this `cache_account()` function, the `is_wp_error()` will pass because the Stripe error object is not a `WP_Error` object. 
image

As a result, the Stripe error object is cached as `$account` and it will be used by code afterwards.

Should we check something like `if ( is_wp_error( $account ) && is_stripe_error($account))`?

Comment (Mayisha) [reply includes/class-wc-stripe-account.php:83]: @harriswong nice catch :100: I missed this case previously. 

What do you think about handling this in the `retrieve` method itself? 

Change [this condition](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-api.php#L180) to the following and throw an exception? 
`if ( is_wp_error( $response ) || empty( $response['body'] ) || isset( $response['body']['message'] ) )`

Comment (harriswong) [reply includes/class-wc-stripe-account.php:83]: I think it is a good idea to check for errors in the `retrieve` function. Though, I would be a bit cautious when changing this behaviour. Currently, we return the `$response['body']` regardless of its HTTP statuses. We sometimes leave it to the function caller to deal with the error handling ([example](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/f5f9e87123a22c4c54e28f3a06e0c1af47de20be/includes/class-wc-stripe-customer.php#L300-L304)). If we throw an error before returning the Stripe's Error Object (`json_decode($response['body'])`), then existing error handling may be affected.

If we want to catch errors on the API response level, then I recommend we check all existing calls to `WC_Stripe_Customer::retrieve()` to make sure nothing breaks if a `WP_Error` is thrown.

> Change this condition to the following and throw an exception?
if ( is_wp_error( $response ) || empty( $response['body'] ) || isset( $response['body']['message'] ) )

The `$response['body']` is a JSON object from Stripe that contains more info on the error itself. To use it, you will need to do `json_decode()` first. Alternatively, we can also make use of the HTTP response code from `$response['code']`. We can probably throw errors for all non-200 responses, according to the [Stripe error doc](https://stripe.com/docs/api/errors).

Comment (harriswong) [reply includes/class-wc-stripe-account.php:83]: Does this block the 5.6 release? If so, I vote for patching the error notice, then refactoring the error handling in a follow-up issue.

Comment (Mayisha) [reply includes/class-wc-stripe-account.php:83]: I have made the hotfix in [this commit](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1969/commits/d596c06711cd3363cf8c5b1002ffbe139320c4dd). It handles the case you mentioned in [this commit](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1969#discussion_r716895364).

**Review by harriswong — DISMISSED** [review]
The latest change stops a Stripe error object to return as a valid `$account` object. Instead, it returns an empty `[]` similar to the existing error handling.
Confirmed errors no longer surfaced. 👍 

** edited **
After merging latest `develop`, I noticed that the `includes/class-wc-stripe-account.php` line 79 and line 83 are duplicated.

**Comment by Mayisha** [comment]
I think `git rebase` messed up when updating the branch. Removed duplicate block.

**Review by waclawjacek — APPROVED** [review]
Great stuff! :shipit:

**Pull request merged**
